static/js/event-list.js

```diff
@@ -0,0 +1,190 @@
+
+    // Build the list in descending order by date.
+    for (const event of events) {
+      this.prependEvent(event);
+    }
+
+    // Now, set up a websocket to listen for new events.
+    this.ws = new WebSocket(`${this.wsHref}?jwt=${API.jwt}`);
+    this.ws.addEventListener('open', () => {
+      if (description.hasOwnProperty('events')) {
+        const msg = {
+          messageType: 'addEventSubscription',
+          data: {},
+        };
+
+        for (const name in description.events) {
+          msg.data[name] = {};
+        }
+
+        this.ws.send(JSON.stringify(msg));
+      }
+    });
+
+    // When a new event comes in, prepend it to the list.
+    this.ws.addEventListener('message', (event) => {
+      const message = JSON.parse(event.data);
+      if (message.messageType === 'event') {
+        const events = Object.keys(message.data).map((name) => {
+          const timestamp = new Date(message.data[name].timestamp);
+          return Object.assign(message.data[name], {name, timestamp});
+        }).sort((a, b) => a.timestamp - b.timestamp);
+
+        for (const event of events) {
+          this.prependEvent(event);
+        }
+      }
+    });
+  }).catch((e) => {
+    console.error(`Error fetching events: ${e}`);
+  });
+
+  // Update event timestamps every 10 seconds.
+  setInterval(() => this.updateTimestamps(), 10000);
+};
+
+EventList.prototype.render = function() {
+  const element = document.createElement('div');
+  const list = '<ul class="event-list"></ul>';
+  element.innerHTML = list;
+  return this.container.appendChild(element.firstChild);
+};
+
+EventList.prototype.prependEvent = function(event) {
+  if (!this.schema.hasOwnProperty(event.name)) {
+    return;
+  }
+
+  const schema = this.schema[event.name];
+
+  let data = event.data;
+  if (typeof data !== 'object' || Array.isArray(data)) {
+    data = {data};
+  }
+
+  let body = '';
+  for (const name of Array.from(Object.keys(data)).sort()) {
+    if (!schema.hasOwnProperty(name)) {
+      continue;
+    }
+
+    let value = data[name];
+
+    switch (schema[name].type) {
+      case 'number':
+      case 'integer':
+        body += `${Utils.escapeHtml(name)}: ${Utils.escapeHtml(value)}`;
+        if (schema[name].hasOwnProperty('unit')) {
+          body += ` ${schema[name].unit}`;
+        }
+        break;
+      case 'object':
+      case 'array':
+        value = JSON.stringify(value);
+        // eslint-ignore-next-line no-fallthrough
+      case 'boolean':
+      case 'string':
+      default:
+        body += `${Utils.escapeHtml(name)}: ${Utils.escapeHtml(value)}`;
+        break;
+    }
+
+    body += '<br>';
+  }
+
+  const el = document.createElement('div');
+
+  const item = `<li class="event-item">
+    <div class="event-header">
+      <span class="event-name">${Utils.escapeHtml(event.name)}</span>
+      <span class="event-time" original="${event.timestamp.toISOString()}"
```

Comment:
It would be more standard to use 'data-original' here and el.dataset.original later